Add CloneCache

mpowelson opened the pull request:
@Levi-Armstrong See what you think about this. I took your tesseract_ros code and made it more general. It is not complete, but the idea should be there. Anything that needs to be cached would need these methods

```
    T::Ptr clone() const
    int getRevision() const;
    bool update(T::Ptr&);  // optional
```

Levi-Armstrong commented on tesseract/tesseract_common/include/tesseract_common/clone_cache.h:
```diff
@@ -0,0 +1,159 @@
+
+    // Lock cache
+    std::unique_lock lock(cache_mutex_);
+    cache_.push_back(clone);
```
this could be null maybe change the above to 
``` c++
if (original_ == nullptr)
  return;

clone = original_->clone();
```

Levi-Armstrong commented on tesseract/tesseract_common/include/tesseract_common/clone_cache.h:
```diff
@@ -0,0 +1,159 @@
+    std::unique_lock<std::shared_mutex> lock(cache_mutex_);
+    std::shared_ptr<T> thor;
+    {
+      auto lock = original_->getMutex();
```
I could see this being need in case where the original and cache are being used in a multi threaded application but  it maybe better for the developer to handle this outside of this class keeping the cache class simpler where it only worries about its internal data being access in a multi threaded case. What do you think?

mpowelson replied:
Yeah that makes sense. The goal was to make a CloneCache<T> object be somewhat interchangeable with a T object. So the user would be responsible for locking it when changing original_

Levi-Armstrong reviewed (commented):
Overall I like it and thank you for taking the time to put this together. It looks like the only requirement on the class is that it has a clone() and getRevision() method if the getMutex is removed. Do you plan to update the planning server to use this?

mpowelson commented:
> Do you plan to update the planning server to use this?

If you think that would be best we can. I was mostly thinking that it would be good to be able to cache other things like contact managers or kinematics. 

Have you seen a speed improvement in the planning server? I would think you should have.

Levi-Armstrong commented:
> > Do you plan to update the planning server to use this?
> 
> If you think that would be best we can. I was mostly thinking that it would be good to be able to cache other things like contact managers or kinematics.
> 
> Have you seen a speed improvement in the planning server? I would think you should have.

Well the version before was using the main Tesseract and not cloning so I believe the performance may not change, but the new approach is more robust.

mpowelson commented:
@Levi-Armstrong I think this is ready for review. I'd be interested to see what your opinion of `const std::shared_ptr<T>& operator->() { return original_; }` is. 

The danger is that if you make a CloneCache::Ptr then you might think you were using the original when you were actually using the CloneCache. If you just say that you always make CloneCache objects instead of ptrs then it acts the way I intended. However, then you `->clone` would call T::clone() and `.clone()` would call CloneCache::clone().

mpowelson commented:
I just changed shared_mutex to a mutex to avoid C++17. I don't think the shared got us much anyway

Levi-Armstrong commented:
If you get the SFINAE working I would also like to add this to the command language.

mpowelson commented:
I'm not sure if it is possible with the templates, but regardless I won't get to look into this soon. You can review it.

mpowelson commented:
@Levi-Armstrong Is it possible with #453 to check if there is a method that takes a certain argument? At least the way it is now, CloneCache update takes an argument. `bool CloneCache::update(T::Ptr&);`

Levi-Armstrong reviewed (commented):
> @Levi-Armstrong Is it possible with #453 to check if there is a method that takes a certain argument? At least the way it is now, CloneCache update takes an argument. `bool CloneCache::update(T::Ptr&);`

I think if this check `CREATE_MEMBER_FUNC_SIG_CHECK` [here](https://stackoverflow.com/questions/257288/templated-check-for-the-existence-of-a-class-member-function) is added it may provide what you are looking for.

``` c++
//Func signature MUST have T as template variable here... simpler this way :\
CREATE_MEMBER_FUNC_SIG_CHECK(x, void (T::*)(), void__x);
bool has_func_sig_void__x = has_member_func_void__x<class_to_check_for_x>::value;
```

mpowelson commented on tesseract/tesseract_common/include/tesseract_common/clone_cache.h:
```diff
@@ -0,0 +1,193 @@
+
+namespace tesseract_common
+{
+///** @brief SFINAE test for whether or not T has an update method */
+// template <typename T>
+// class SupportsUpdate
+//{
+//  typedef char one;
+//  struct two
+//  {
+//    char x[2];
+//  };
+
+//  template <typename C>
+//  static one test(decltype(&C::update));
+//  template <typename C>
+//  static two test(...);
+
+// public:
+//  enum
+//  {
+//    value = sizeof(test<T>(0)) == sizeof(char)
+//  };
+//};
```
Ok. If you want to just push to this branch, feel free. Or I can update it later.

Levi-Armstrong replied:
Sure, I will add the check and a unit test to tesseract_common. Then try adding it here.

Levi-Armstrong replied:
@mpowelson I tried the one in the link which did not do exactly what we wanted so after a lot of testing and searching I was able to come up with a solution that works. The additional checks have been added in PR #454.

mpowelson commented:
I rebased it but have not added the sfinae checks

mpowelson commented:
@Levi-Armstrong This has the sfinae checks now and should be ready for review

mpowelson commented:
@Levi-Armstrong Just a reminder that this is ready for review

Levi-Armstrong commented on tesseract/tesseract_common/include/tesseract_common/clone_cache.h:
```diff
@@ -0,0 +1,185 @@
+    , cache_size_(static_cast<std::size_t>(cache_size))
+
+  {
+    for (long i = 0; i < cache_size; i++)
```
Missing the static assert here using the function generated by `CREATE_MEMBER_FUNC_SIGNATURE_CHECK`.

mpowelson replied:
I should be checking it everywhere that it is used. The idea is that it is optional to support updating.

Levi-Armstrong replied:
So what does the update function do?

mpowelson replied:
It updates the object to be at the state of the pattern passed in. So in the case of the environment it would apply any additional commands that have been applied on the environment that's passed in

mpowelson replied:
The idea is that we may want caches of things that don't keep a history and can't be updated easily, so for those if they're out of date the only thing you can do is reclone them

Levi-Armstrong reviewed (commented):
Other than the one comment. Looks good.

mpowelson commented:
Add sfinae checks for getRevision and clone

mpowelson commented:
@Levi-Armstrong Added the sfinae checks for getRevision and clone. Also rebased onto latest.

Levi-Armstrong commented:
It looks like the CI is failing because tesseract_common uses console_bridge but it is no in the package.xml or cmake file. THis looks like it has always been missing. Changes highlighted below should solve the issue and also need to add target `console_bridge::console_bridge` to `target_link_libraries`.

Add to  CMakeLists.txt:
``` cmake
find_package(console_bridge REQUIRED)

if(NOT TARGET console_bridge::console_bridge)
  add_library(console_bridge::console_bridge INTERFACE IMPORTED)
  set_target_properties(console_bridge::console_bridge PROPERTIES INTERFACE_INCLUDE_DIRECTORIES ${console_bridge_INCLUDE_DIRS})
  set_target_properties(console_bridge::console_bridge PROPERTIES INTERFACE_LINK_LIBRARIES ${console_bridge_LIBRARIES})
else()
  get_target_property(CHECK_INCLUDE_DIRECTORIES console_bridge::console_bridge INTERFACE_INCLUDE_DIRECTORIES)
  if (NOT ${CHECK_INCLUDE_DIRECTORIES})
    set_target_properties(console_bridge::console_bridge PROPERTIES INTERFACE_INCLUDE_DIRECTORIES ${console_bridge_INCLUDE_DIRS})
  endif()
endif()
```

Add to tesseract_common-config.cmake.in:

``` cmake
find_dependency(console_bridge REQUIRED)

if(NOT TARGET console_bridge::console_bridge)
  add_library(console_bridge::console_bridge INTERFACE IMPORTED)
  set_target_properties(console_bridge::console_bridge PROPERTIES INTERFACE_INCLUDE_DIRECTORIES ${console_bridge_INCLUDE_DIRS})
  set_target_properties(console_bridge::console_bridge PROPERTIES INTERFACE_LINK_LIBRARIES ${console_bridge_LIBRARIES})
else()
  get_target_property(CHECK_INCLUDE_DIRECTORIES console_bridge::console_bridge INTERFACE_INCLUDE_DIRECTORIES)
  if (NOT ${CHECK_INCLUDE_DIRECTORIES})
    set_target_properties(console_bridge::console_bridge PROPERTIES INTERFACE_INCLUDE_DIRECTORIES ${console_bridge_INCLUDE_DIRS})
  endif()
endif()
```

mpowelson commented:
@Levi-Armstrong I added that. Why do we redefine the target? Doesn't console bridge already have one that is just `console_bridge`? Regardless, I added the lines like you said as a separate commit. If you want, we can squash it

Levi-Armstrong commented:
> @Levi-Armstrong I added that. Why do we redefine the target? Doesn't console bridge already have one that is just `console_bridge`? Regardless, I added the lines like you said as a separate commit. If you want, we can squash it

They changed the target name between Bionic and Noetic. Before Noetic the target was `console_bridge` and for Noetic it is `console_bridge::console_bridge` which is the recommended format. This makes Tesseract compatible with Noetic along with the older distributions.

mpowelson commented:
@Levi-Armstrong It is finally passing

Levi-Armstrong commented on tesseract/tesseract_common/cmake/tesseract_common-config.cmake.in:
```diff
@@ -15,5 +15,18 @@ else()
     find_dependency(Boost COMPONENTS system filesystem program_options)
 endif()
 
+find_dependency(console_bridge REQUIRED)
```
Required can be removed.

mpowelson replied:
done

Levi-Armstrong reviewed (changes requested):
One minor change. Then good to merge.
